File: app/src/main/java/de/xikolo/views/CustomFontTextView.java
package de.xikolo.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.util.Log;
import android.widget.TextView;

import java.util.Hashtable;

import de.xikolo.R;
import de.xikolo.config.Config;

public class CustomFontTextView extends TextView {

    private static final String TAG = CustomFontTextView.class.getSimpleName();

    private static final Hashtable<String, Typeface> cache = new Hashtable<>();

    public CustomFontTextView(Context context) {
        super(context);
    }

    public CustomFontTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setCustomFont(context, attrs);
    }

    public CustomFontTextView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        setCustomFont(context, attrs);
    }

    private void setCustomFont(Context ctx, AttributeSet attrs) {
        TypedArray a = ctx.obtainStyledAttributes(attrs, R.styleable.CustomFontTextView);
        String customFont = a.getString(R.styleable.CustomFontTextView_customFont);
        setCustomFont(ctx, customFont);
        a.recycle();
    }

    public boolean setCustomFont(Context ctx, String asset) {
        Typeface tf;
        synchronized (cache) {
            try {
                if (!cache.containsKey(asset)) {
                    tf = Typeface.createFromAsset(ctx.getAssets(), Config.FONT + asset);
                    cache.put(asset, tf);
                }
                tf = cache.get(asset);
                setTypeface(tf);
            } catch (Exception e) {
                Log.e(TAG, "Could not get typeface: " + e.getMessage());
                return false;
            }
            return true;
        }

    }

}

File: app/src/main/java/de/xikolo/views/CustomFontTextView.kt
package de.xikolo.views

import android.content.Context
import android.graphics.Typeface
import android.util.AttributeSet
import android.util.Log
import androidx.appcompat.widget.AppCompatTextView
import de.xikolo.R
import de.xikolo.config.Config
import java.util.*

class CustomFontTextView : AppCompatTextView {

    companion object {
        private val TAG = CustomFontTextView::class.java.simpleName
        private val cache = Hashtable<String, Typeface>()
    }

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        setCustomFont(context, attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        setCustomFont(context, attrs)
    }

    private fun setCustomFont(ctx: Context, attrs: AttributeSet) {
        val a = ctx.obtainStyledAttributes(attrs, R.styleable.CustomFontTextView)
        val customFont = a.getString(R.styleable.CustomFontTextView_customFont)
        setCustomFont(ctx, customFont)
        a.recycle()
    }

    fun setCustomFont(ctx: Context, asset: String?): Boolean {
        var tf: Typeface?
        synchronized(cache) {
            try {
                if (!cache.containsKey(asset)) {
                    tf = Typeface.createFromAsset(ctx.assets, Config.FONT_DIR + asset!!)
                    cache[asset] = tf
                }
                tf = cache[asset]
                typeface = tf
            } catch (e: Exception) {
                Log.e(TAG, "Could not get typeface: " + e.message)
                return false
            }

            return true
        }

    }

}
